File: rsa.py
import random
import math


def gcd(a, b):
    if b == 0:
        return a
    return gcd(b, a % b)


def prime_finder(n):
    test_number = random.randrange(n, n+50)
    for i in range(2, test_number):
        if (test_number % i) == 0:
            return prime_finder(n)
    return test_number


def pub_key1(phi):
    pub_keys = []
    for i in range(2, phi):
        if gcd(i, phi) == 1:
            pub_keys.append(i)
        if len(pub_keys) >= 50:
            break
    e = random.choice(pub_keys)
    del(pub_keys)
    return e


def gcd2(p, q):

    while(q):
        p, q = q, p % q

    return p


def finder2(n):
    test_number = random.randrange(n+1, n+50, 2)
    limit = math.floor((math.sqrt(test_number)))
    for i in range(3, limit+1, 2):
        if (test_number % i) == 0:
            return finder2(n)
    return test_number


def pub_key(phi):
    pub_keys = []
    for i in range(2, phi):
        if gcd2(i, phi) == 1:
            pub_keys.append(i)
        if len(pub_keys) >= 50:
            break
    e = random.choice(pub_keys)
    del(pub_keys)
    return e


def priv_key(e, phi):
    i = 2
    while 1:
        if (i * e) % phi == 1:
            d = i
            break
        i += 1
    return d

File: rsa.c
#include <math.h>
#include <stdio.h>
#include <time.h>
#include <stdlib.h>

int gcd2(int a, int b)
{
    int nuevo_a = 0;
    while (b)
    {
        nuevo_a = a;
        a = b;
        b = nuevo_a % b;
    }
    return a;
}

void inicia()
{
    srand(time(NULL));
}

int prime_finder2(int n)
{
    int test_number = rand() % (51) + (n);
    int i;
    if (test_number % 2 == 0)
    {
        return prime_finder2(n);
    }
    else
    {
        int a = (int)sqrt(test_number);
        for (i = 3; i < a + 1; i += 2)
        {
            if (test_number % i == 0)
            {
                return prime_finder2(n);
            }
        }
        return test_number;
    }
}

int pubkeys(int phi)
{
    int i;
    int j = 0;
    int N = 50;
    int pub_keys[N];
    for (i = 0; i < (sizeof(pub_keys) / sizeof(pub_keys[0])); i++)
    {
        pub_keys[i] = 0;
    }
    for (i = 2; i < phi; i++)
    {
        if (gcd2(i, phi) == 1)
        {
            pub_keys[j] = i;
            j++;
        }
        if (j > (N - 1))
        {
            break;
        }
    }
    int e;
    while (1)
    {
        e = pub_keys[rand() % (sizeof(pub_keys) / sizeof(pub_keys[0]))];
        if (e == 0)
        {
            continue;
        }
        else
        {
            break;
        }
    }
    for (i = 0; i < (sizeof(pub_keys) / sizeof(pub_keys[0])); i++)
    {
        pub_keys[i] = 0;
    }
    return e;
}

int privkeys(int e, int phi)
{
    int i = 2;
    int d = 0;
    while (1)
    {
        if ((i * e) % phi == 1)
        {
            d = i;
            break;
        }
        i += 1;
    }
    return d;
}
